File: Lab1/graph.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main()
{
    int n, i;
    int fairHeads = 0, biasedHeads = 0;

    printf("Enter the number of coin tosses: ");
    scanf("%d", &n);

    FILE *fp = fopen("coin_data.txt", "w");

    if(fp == NULL)
    {
        printf("Error opening file!\n");
        return 1;
    }

    fprintf(fp, "Toss Fair_Probability Biased_Probability\n");

    srand(time(NULL));

    for(i = 1; i <= n; i++)
    {
        // Fair coin
        if(rand() % 2 == 0)
        {
            fairHeads++;
        }

        // Biased coin (70% heads)
        double r = (double)rand() / RAND_MAX;
        if(r < 0.7)
        {
            biasedHeads++;
        }

        fprintf(fp, "%d %.6f %.6f\n",
                i,
                (double)fairHeads / i,
                (double)biasedHeads / i);
    }

    fclose(fp);

    printf("Final Results:\n");
    printf("Fair Coin Probability = %.6f\n", (double)fairHeads / n);
    printf("Biased Coin Probability = %.6f\n", (double)biasedHeads / n);

    printf("\nData has been saved to coin_data.txt\n");

    return 0;
}

File: Lab1/graph.py
import matplotlib.pyplot as plt

toss = []
fair = []
biased = []

with open("coin_data.txt") as f:
    next(f)  # Skip header

    for line in f:
        t, f_prob, b_prob = line.split()
        toss.append(int(t))
        fair.append(float(f_prob))
        biased.append(float(b_prob))

plt.plot(toss, fair, label="Fair Coin")
plt.plot(toss, biased, label="Biased Coin (70%)")

plt.axhline(0.5, linestyle='--', label='0.5')
plt.axhline(0.7, linestyle='--', label='0.7')

plt.xlabel("Number of Tosses")
plt.ylabel("Estimated Probability of Heads")
plt.title("Law of Large Numbers")
plt.legend()
plt.grid(True)
plt.show()